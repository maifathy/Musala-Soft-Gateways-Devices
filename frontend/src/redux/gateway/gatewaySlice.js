import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  gateways: []
};
const gatewaysStr = 'gateways';
export const gatewaySlice = createSlice({
  name: 'gateway',
  initialState,
  reducers: {
    setGateways: (state, action) => {
      const updatedState = { ...state };
      updatedState[gatewaysStr] = action.payload;
      return updatedState;
    },
    removeGateway: (state, action) => {
      state.gateways.filter((gateway) => gateway.id !== action.payload);
    },
    appendGateway: (state, action) => {
      state.gateways.push(action.payload);
    },
    removeDevice: (state, action) => {
      state.gateways.filter((gateway) => {
        return gateway.devices.map((device, index) => {
          if (device._id === action.payload) {
            gateway.devices.splice(index, 1);
          }
          return gateway.devices;
        });
      });
    },
    appendDevice: (state, action) => {
      state.gateways.filter((gateway) => {
        if (gateway._id === action.payload.gatewayId) {
          gateway.devices.push(action.payload.device);
        }
        return gateway.devices;
      });
    }
  }
});

// Action creators are generated for each case reducer function
export const {
  setGateways, removeGateway, appendGateway, removeDevice, appendDevice
} = gatewaySlice.actions;

export default gatewaySlice.reducer;
