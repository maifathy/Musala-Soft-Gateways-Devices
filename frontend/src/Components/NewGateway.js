import React, { useEffect, useState, useRef, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { appendGateway } from '../redux/gateway/gatewaySlice.js';
import { addGateway } from '../utils/api.js';
import styled from 'styled-components';

const NewGateway = () => {
  const [name, setName] = useState('');
  const [ip, setIP] = useState('');
  const [message, setMessage] = useState('');
  const dispatch = useDispatch();
  const inputIP = useRef(null);
  const isUnmountedRef = useRef(false);

  useEffect(() => {
    return () => {
      // cleanup function will occur when component unmounted.
      isUnmountedRef.current = true;
    } 
  },[]);

  useEffect(() => {
    if(message === '')
      return;
    setTimeout(() => setMessage(''), 7000);
  }, [message]);

  const validateIPaddress = () => {
    const format = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
    if (inputIP.current.value.match(format)) {
      setMessage('');
      return true;
    }
    inputIP.current.focus();
    setMessage('You entered an invalid IP address!!');
    return false;
  };

  const newGateway = useCallback((e) => {
    setMessage('');
    e.preventDefault();
    if (validateIPaddress()) {
      addGateway(name, ip)
        .then((obj) => {
          if (obj.status === 200) {
            dispatch(appendGateway(obj.gateway));
          }
          if(!isUnmountedRef.current) setMessage(obj.message);
        })
        .catch((err) => { setMessage(err.message); });
      if(!isUnmountedRef.current) {
        setName('');
        setIP('');
      }
    }
  }, [name, ip]);

  return (
    <Wrapper>
      <h3>New Gateway</h3>
      <form onSubmit={ newGateway } data-testid="form">
          <Input type='text'  data-testid="name_input" value={name} placeholder='Name' onChange={(e) => setName(e.target.value)}/>
        |
          <Input type='text' data-testid="ip_input" value={ip} placeholder='IP address' ref={inputIP} onChange={(e) => setIP(e.target.value)}/>
        <div>
          <Button type='submit' disabled={name === '' || ip === ''}>Submit</Button>
        </div>
      </form>
      { message && <p data-testid="message">{message}</p> }
    </Wrapper>
  );
};

export default NewGateway;

const Wrapper = styled.div`
  width: 500px;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 50px;
  margin-bottom: 50px;
  padding: 20px;
  box-shadow: 0 0 11px rgb(43 52 58 / 50%)  ; 
  -webkit-box-shadow: 0 0 11px rgb(43 52 58 / 50%)  ; 
  -moz-box-shadow: 0 0 11px rgb(43 52 58 / 50%)  ; 
`;
const Input = styled.input`
  border: 1px solid #a6a6ca;
  border-radius: 2px;
  font-size: 14px;
  color: navy;
  height: 30px;
  :focus {
    outline: none;
    background-color: rgb(215 215 229);
  }
`;

const Button = styled.button`
  margin-top: 10px;
  padding: 5px;
  width: 200px;
  border: 1px solid "#7620ff";
  background-color: "#7620ff";
  outline: none;
  color: "#fff";
  :hover {
    background-color: "#352878";
  }
`;