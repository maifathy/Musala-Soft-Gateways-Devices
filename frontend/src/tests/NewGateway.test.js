/* eslint-disable no-undef */
import React from 'react';
import Enzyme,{ mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { render as renderLibrary, fireEvent, cleanup, act } from '@testing-library/react';
import { waitFor } from "@testing-library/dom";
import { store } from '../app/store.js';
import { initialState } from '../redux/gateway/gatewaySlice.js';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import history from '../utils/history.js';
import NewGateway from '../Components/NewGateway.js';
import { clearCollections } from '../utils/api.js';

Enzyme.configure({ adapter: new Adapter() })

it('New Category Snapshot - renders correctly enzyme', () => {
  const wrapper = mount(<Provider store={store}>  <BrowserRouter history={history} location={history.location} navigator={history}> <NewGateway /> </BrowserRouter> </Provider>);

  expect(toJson(wrapper)).toMatchSnapshot();
});

describe('Testing state, first load', () => {
  it('should return the initial state', () => {
    expect(initialState)
    .toEqual({ gateways: [] })
  })
});

afterEach(() => { 
  cleanup(); 
  clearCollections().then(() => {
    mongoose.connection.close();
  });
});

// Invalid IP address!!
it('Invalid IP address!!', async () => {
  const { getByText, getByTestId } = renderLibrary(<Provider store={store}>  <BrowserRouter history={history} location={history.location} navigator={history}> <NewGateway /> </BrowserRouter> </Provider>);

  expect(getByText('Submit')).toBeDisabled();

  fireEvent.change(getByTestId('name_input'), { target: { value: 'any name'}});
  fireEvent.change(getByTestId('ip_input'), { target: { value: 'abc'}});

  expect(getByText('Submit')).not.toBeDisabled();

  fireEvent.submit(getByTestId("form")); 
  //fireEvent.click(getByText('Submit')); // alternative Submit

  await waitFor(() =>
    expect(getByTestId('message')).toHaveTextContent('You entered an invalid IP address!!')
  );
});

// Valid Gateway details!!
it('Valid Gateway details!!', async () => {
  const { getByText, getByTestId } = renderLibrary(<Provider store={store}>  <BrowserRouter history={history} location={history.location} navigator={history}> <NewGateway /> </BrowserRouter> </Provider>);

  expect(getByText('Submit')).toBeDisabled();

  fireEvent.change(getByTestId('name_input'), { target: { value: 'Gateway1'}});
  fireEvent.change(getByTestId('ip_input'), { target: { value: '127.0.0.1'}});

  expect(getByText('Submit')).not.toBeDisabled();

  act(() => {
    fireEvent.submit(getByTestId("form")); 
    //fireEvent.click(getByText('Submit')); // alternative Submit

    waitFor(() =>
      expect(getByTestId('message')).toHaveTextContent('Gateway is successfully added'),
      expect(getByTestId('name_input')).toBeEmptyDOMElement(),
      expect(getByTestId('ip_input')).toBeEmptyDOMElement()
    ); 
  })
});

