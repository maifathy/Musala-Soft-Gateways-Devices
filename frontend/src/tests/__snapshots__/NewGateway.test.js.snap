// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Category Snapshot - renders correctly enzyme 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
    
  <BrowserRouter
    history={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigator={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Router
      location={
        Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        Object {
          "action": "POP",
          "createHref": [Function],
          "createURL": [Function],
          "encodeLocation": [Function],
          "go": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
       
      <NewGateway>
        <styled.div>
          <div
            className="sc-bgqQcB dxpCdu"
          >
            <h3>
              New Gateway
            </h3>
            <form
              data-testid="form"
              onSubmit={[Function]}
            >
              <styled.input
                data-testid="name_input"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
              >
                <input
                  className="sc-gTRrQi kOdMJF"
                  data-testid="name_input"
                  onChange={[Function]}
                  placeholder="Name"
                  type="text"
                  value=""
                />
              </styled.input>
              |
              <styled.input
                data-testid="ip_input"
                onChange={[Function]}
                placeholder="IP address"
                type="text"
                value=""
              >
                <input
                  className="sc-gTRrQi kOdMJF"
                  data-testid="ip_input"
                  onChange={[Function]}
                  placeholder="IP address"
                  type="text"
                  value=""
                />
              </styled.input>
              <div>
                <styled.button
                  disabled={true}
                  type="submit"
                >
                  <button
                    className="sc-ewnqHT gbhdnH"
                    disabled={true}
                    type="submit"
                  >
                    Submit
                  </button>
                </styled.button>
              </div>
            </form>
          </div>
        </styled.div>
      </NewGateway>
       
    </Router>
  </BrowserRouter>
   
</Provider>
`;
